code/controllers/subsystem/html_audio.dm

```diff
@@ -0,0 +1,274 @@
+SUBSYSTEM_DEF(html_audio)
+	name = "HTML Audio"
+	init_order = INIT_ORDER_HTMLAUDIO
+	flags = SS_NO_FIRE
+	var/max_channels = 512 // set this to how many total channels you want
+	var/list/channel_assignment = list() // list([atom in the world], [atom in the world], ...) for keeping track of what channels are in use and by what
+	var/list/channel_loop_status = list() // list([atom in the world] = TRUE/FALSE, ...) for keeping track of what channels are looping
+	var/list/channel_tts_status = list() // list([atom in the world] = TRUE/FALSE, ...) for keeping track of what channels are TTS
+	var/list/listeners = list() // list of client listeners to update when audio gets added
+	var/list/active_urls = list() // list("url_here", ...), ref'd by active_urls[channel_id]
+	var/list/list/channel_requires_LOS_at_start_listeners = list() // list(list(/client, ...)) complicated, used for LOS requirements on hearing speech
+	var/list/channel_requires_LOS_at_start = list() // list([atom in the world] = TRUE/FALSE, ...) for keeping track of what channels require LOS at the start
+	var/list/listener_handlers = list() // list([atom in the world] = listener_handler)
+	var/browse_txt
+	var/preview_browse_txt
+
+/datum/controller/subsystem/html_audio/Initialize()
+	browse_txt = {"
+			<META http-equiv="X-UA-Compatible" content="IE=edge">"}
+	for(var/i in 1 to max_channels)
+		browse_txt += {"
+			<audio id="channel_[i]" volume="0">
+			</audio>"}
+	browse_txt += {"
+			<script>
+			function setVolume(volume, element_id)
+			{
+				var audio_player = document.getElementById(element_id);
+				audio_player.volume = parseFloat(volume);
+			}
+			function playAudio(url, element_id)
+			{
+				var audio_player = document.getElementById(element_id);
+				audio_player.pause();
+				audio_player.src = url;
+				audio_player.load();
+				audio_player.play();
+			}
+			function setLooping(loop, element_id)
+			{
+				var audio_player = document.getElementById(element_id);
+				var isTrueSet = (loop === 'true');
+				if(isTrueSet)
+				{
+					audio_player.loop = true;
+				}
+				else
+				{
+					audio_player.loop = false;
+				}
+			}
+			</script>
+	"}
+	preview_browse_txt = {"
+			<META http-equiv="X-UA-Compatible" content="IE=edge">
+			<audio id="channel_preview" volume="100">
+			</audio>
+			<script>
+			function playAudio(url)
+			{
+				var audio_player = document.getElementById("channel_preview");
+				audio_player.pause();
+				audio_player.src = url;
+				audio_player.load();
+				audio_player.play();
+			}
+			</script>
+	"}
+	channel_assignment.len = max_channels
+	channel_loop_status.len = max_channels
+	channel_tts_status.len = max_channels
+	active_urls.len = max_channels
+	channel_requires_LOS_at_start.len = max_channels
+	channel_requires_LOS_at_start_listeners.len = max_channels
+	RegisterSignal(SSdcs, COMSIG_GLOB_MOB_LOGGED_IN, PROC_REF(register_listener))
```

Comment:
what about people who join in before the subsystem inits?

this only works now because its only consumer (tts) doesn't send to lobby mobs, atm only lobby mobs who connected after this subsystem init'ed will get a the html window sent to them, the rest won't get it until their next mob change